[Core] make per-cloud catalog lookup parallel

aylei opened the pull request:
<!-- Describe the changes in this PR -->

Part of https://github.com/skypilot-org/skypilot/issues/3159

This is a **draft PR** with a known issue to be solved.

The test is done locally with the duration of launch to confirm timed (https://github.com/aylei/skypilot/commit/e662df00ba072f7ca81a02889f4725d65d4d30b3). On my laptop, this PR improves the performance in 4 test cases with 7 clouds enabled compared to main (https://github.com/skypilot-org/skypilot/commit/f0ebf13b5b164d012f4b033609eb1b364dc6caf3):

1. `sky launch --gpus H100:8`, from 2.38s to 2.06s
2. `rm -rf ~/.sky/catalogs/** && sky launch --gpus H100:8` (cold start), from 12.15s to 6.29s
3. `sky launch test.yaml`, ordered resources with non-canonical names, from 5.75s to 5.07s
4. `rm -rf ~/.sky/catalogs/** && sky launch test.yaml`, same as above, but cold start, from 15.75s to 10.17s

test.yaml:

```
resources:
  cpus: 8+
  ordered:
    - accelerators: L40S:1
    - accelerators: L40:1
    - accelerators: A10g:1
    - accelerators: A10:1
    - accelerators: A100:1
    - accelerators: H100:1

run: |
  echo 'test done'
```

Cold-start and test.yaml are chosen deliberately. Because I found that for the case of "--gpu H100:8", the improvement was less than I expected (I expected a ~1s improvement according to [former analysis](https://github.com/skypilot-org/skypilot/issues/3159#issuecomment-2548747327)). 

The reason is GIL. Through profiling (source: [gil.svg](https://github.com/user-attachments/assets/e71136c3-4742-4e5e-b136-0dccb534ce42)), I found that each cloud spends 10~100+ ms on init, which can not be parallelized due to GIL:

<img width="1496" alt="image" src="https://github.com/user-attachments/assets/e9470d4a-26ed-46d4-8ff4-3f4a58f5e3a5" />

Actually in my setup only K8S and AWS would file network IO, where K8S ListNode costs ~0.5s and AWS get-caller-identity costs ~1s. So I tested the cold-start case which shows in scenarios with more I/O calls, the performance optimization of concurrency will be more significant.

For `test.yaml`, I deliberately included non-canonical GPU names to cover the non short-circuit path, where the IO calls will be advanced to `get_accelerators`.

I thought carefully about whether this PR is a premature optimization and my final thought is that processing each cloud concurrently in multi-cloud scenario is necessary for scalability as skypilot is a multi-cloud solution at its core(IIUC). This is why I did not parallelize the single `get_accelerators` function but parallelize all catalog lookup calls in `_map_clouds_catalog`. @Michaelvll What do you think?

Note, since sub thread status is silent yet, sky launch will lose catalog progress after this PR merged. Follow up: https://github.com/skypilot-org/skypilot/issues/4491

<!-- Describe the tests ran -->
<!-- Unit tests (tests/test_*.py) are part of GitHub CI; below are tests that launch on the cloud. -->

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (see above)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [x] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

Michaelvll commented on sky/utils/subprocess_utils.py:
```diff
@@ -293,3 +293,23 @@ def kill_process_daemon(process_pid: int) -> None:
+    """Apply a function to a list of clouds,
+    with parallelism if there is more than one cloud.
+    """
+    count = len(clouds)
+    if count == 0:
+        return []
+    # Short-circuit in single cloud setup.
+    if count == 1:
+        return [func(clouds[0])]
```
How about we have the `run_in_parallel` to automatically choose whether to short-circuit a thread pool based on the length of the input? In this case, we can avoid a new API, which also benefits the existing code.

aylei replied:
Good call, thanks!

Michaelvll commented on sky/optimizer.py:
```diff
@@ -1293,9 +1305,10 @@ def _fill_in_launchable_resources(
-            feasible_resources = cloud.get_feasible_launchable_resources(
-                resources, num_nodes=task.num_nodes)
+
+        feasible_list = subprocess_utils.maybe_parallelize_cloud_operation(
+            _make_resource_finder(resources, task.num_nodes), clouds_list)
```
wondering if we can simplify this with `run_in_parallel(lambda cloud: cloud.get_feasible_launchable_resources(resources, num_nodes), cloud_list)`

aylei replied:
I tried lambda at the very beginning but got a pylint warning😂:

```
sky/optimizer.py:1311:16: W0640: Cell variable resources defined in loop (cell-var-from-loop)
```

aylei replied:
Just figured out how to avoid potential closure issue in loop when using lambda https://stackoverflow.com/questions/25314547/cell-var-from-loop-warning-from-pylint

Michaelvll replied:
Ahh, good catch. Just wondering if we can make the function a bit simpler.

Michaelvll reviewed (commented):
Thanks @aylei! Adding the parallelization for all cloud methods makes sense to me. Left some minor comments mainly for interfaces.

Investigating AWS and kubernetes latency should be a great next step. IIRC, trying to fetch the mapping for AWS zone names can cause a large latency even when we don't have AWS enabled.

aylei commented:
> Thanks @aylei! Adding the parallelization for all cloud methods makes sense to me. Left some minor comments mainly for interfaces.
> 
> Investigating AWS and kubernetes latency should be a great next step. IIRC, trying to fetch the mapping for AWS zone names can cause a large latency even when we don't have AWS enabled.

@Michaelvll Thanks for your suggestions❤️. I reproduced the AWS zone name issue by cleaning catalog cache, excluding `aws` from `allowed_clouds` in `~/.sky/config.yaml` and run `sky launch --gpus L40:8`, which uses a non-canonical GPU name and will enter the slow path https://github.com/skypilot-org/skypilot/blob/c19e94e35c8c386d8ee7111529210ea0ac650883/sky/utils/accelerator_registry.py#L96. I think service_catalog should also honor `allowed_clouds` configuration but this seems involve semantic change, since `allowed_clouds` is documented as "# This field is used to restrict the clouds that SkyPilot will check and use
# when running `sky check`. ". My idea is to apply the same semantic to other commands like launch, how do you like? @Michaelvll 

As of normal cases, the zone mapping is usually cached with AWS userID as the cache key (https://github.com/skypilot-org/skypilot/blob/c19e94e35c8c386d8ee7111529210ea0ac650883/sky/clouds/service_catalog/aws_catalog.py#L79), so the only relevant RPC  is `get-caller-identity` in daily usage, with the assumption that the AWS account used rarely changes.

Therefore, for the mainstream scenarios, I currently think there might be no optimization method for the `get-caller-identity` RPC of AWS and `ListNode` RPC of K8S without adjusting UX. That is, caching these calls requires extra user attention and should be carefully designed. I will evaluate whether the benefits would payoff.

aylei commented:
@Michaelvll I've addressed the comments and added sub thread status support. Please kindly take a look.

Current look when sub thread status is involved:

![2024-12-19 20 22 20](https://github.com/user-attachments/assets/5ed5302f-6158-42fe-83f7-15b0c830db51)

![2024-12-19 20 23 04](https://github.com/user-attachments/assets/de076b1f-c016-4edb-aa97-82e770d26fbd)

Michaelvll commented on sky/utils/rich_utils.py:
```diff
@@ -1,16 +1,21 @@
 """Rich status spinner utils."""
```
This is great! However, since this can affect many other part of the code, let's move this to another PR, and keep this PR clean for only having the parallelism changes.

aylei replied:
Sure, but since status of sub thread is silent on master, there would be a UX degradation in between, is that okay?

Michaelvll commented on sky/utils/subprocess_utils.py:
```diff
@@ -113,6 +114,12 @@ def run_in_parallel(func: Callable,
       A list of the return values of the function func, in the same order as the
       arguments.
     """
+    if isinstance(args, collections.abc.Sized):
```
nit: how about we change the input args to be `List[Any]` and change the invocations to use lists to avoid this `collections.abc.Sized`?

aylei replied:
Sounds good, I will try this~

Michaelvll reviewed (approved):
Thanks @aylei! It looks good to me. Just some minor nits. Let's proceed with the optimization of the second step. : )

aylei commented on sky/optimizer.py:
```diff
@@ -1293,9 +1293,12 @@ def _fill_in_launchable_resources(
-            feasible_resources = cloud.get_feasible_launchable_resources(
-                resources, num_nodes=task.num_nodes)
+
+        feasible_list = subprocess_utils.run_in_parallel(
+            lambda cloud, r=resources, n=task.num_nodes:
+            (cloud, cloud.get_feasible_launchable_resources(r, n)),
+            clouds_list)
+        for cloud, feasible_resources in feasible_list:
```
Good catch, with this new knowledge, I personally still tend to loosen the behavior dependency on the `run_in_parallel` function for maintainability, wdyt?

aylei commented:
@Michaelvll ready for the next review

Michaelvll commented on sky/clouds/service_catalog/__init__.py:
```diff
@@ -46,7 +46,10 @@ def _map_clouds_catalog(clouds: CloudFilter, method_name: str, *args, **kwargs):
+        return method(*args, **kwargs)
+
+    results = subprocess_utils.run_in_parallel(_execute_catalog_method,
+                                               list(clouds), len(clouds))
```
style nit: use kwargs for non-obvious argument meaning : )

aylei replied:
Thanks!

Michaelvll commented on sky/utils/subprocess_utils.py:
```diff
@@ -113,6 +113,11 @@ def run_in_parallel(func: Callable,
+        return []
+    # Short-circuit for single element
+    if len(args) == 1:
+        return [func(next(iter(args)))]
```
```suggestion
        return [func(args[0])]
```

Michaelvll reviewed (approved):
Thanks @aylei! LGTM. Could you share a bit of how the UX looks like without the thread-based handling for spinner?

aylei commented:
@Michaelvll 

For canonical GPU:
![2024-12-21 21 13 01](https://github.com/user-attachments/assets/ad8924e7-ca2e-45aa-b2b9-069468f352d2)

For non-canonical GPU:
![2024-12-21 21 13 30](https://github.com/user-attachments/assets/07fdf3e3-218c-47b5-9601-460aa3b0c1aa)

Fine when no catalog refresh is involved, otherwise it looks a little confusing since `Optimizing...` or `List accelerators...` would last several seconds with no detailed message.

Michaelvll commented:
> @Michaelvll
> 
> For canonical GPU: ![2024-12-21 21 13 01](https://private-user-images.githubusercontent.com/18556593/397929592-ad8924e7-ca2e-45aa-b2b9-069468f352d2.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY3OTg5NTIsIm5iZiI6MTczNjc5ODY1MiwicGF0aCI6Ii8xODU1NjU5My8zOTc5Mjk1OTItYWQ4OTI0ZTctY2EyZS00NWFhLWIyYjktMDY5NDY4ZjM1MmQyLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEzVDIwMDQxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU4YzBkMjJhNDM5ZmVkNDU4ZTk4MzQ4ZjIxMjMxZDY0YmVjNWEwMGY3ODMxODE1YWZhZjVkNjhjMmQwN2JjNDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.0TdD_MXq4CvlLD790xS__UhV_GSoIfIsyBfXV2PTwn0) [ ](https://private-user-images.githubusercontent.com/18556593/397929592-ad8924e7-ca2e-45aa-b2b9-069468f352d2.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY3OTg5NTIsIm5iZiI6MTczNjc5ODY1MiwicGF0aCI6Ii8xODU1NjU5My8zOTc5Mjk1OTItYWQ4OTI0ZTctY2EyZS00NWFhLWIyYjktMDY5NDY4ZjM1MmQyLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEzVDIwMDQxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU4YzBkMjJhNDM5ZmVkNDU4ZTk4MzQ4ZjIxMjMxZDY0YmVjNWEwMGY3ODMxODE1YWZhZjVkNjhjMmQwN2JjNDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.0TdD_MXq4CvlLD790xS__UhV_GSoIfIsyBfXV2PTwn0) ![2024-12-21 21 13 01](https://private-user-images.githubusercontent.com/18556593/397929592-ad8924e7-ca2e-45aa-b2b9-069468f352d2.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY3OTg5NTIsIm5iZiI6MTczNjc5ODY1MiwicGF0aCI6Ii8xODU1NjU5My8zOTc5Mjk1OTItYWQ4OTI0ZTctY2EyZS00NWFhLWIyYjktMDY5NDY4ZjM1MmQyLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEzVDIwMDQxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU4YzBkMjJhNDM5ZmVkNDU4ZTk4MzQ4ZjIxMjMxZDY0YmVjNWEwMGY3ODMxODE1YWZhZjVkNjhjMmQwN2JjNDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.0TdD_MXq4CvlLD790xS__UhV_GSoIfIsyBfXV2PTwn0) [ ](https://private-user-images.githubusercontent.com/18556593/397929592-ad8924e7-ca2e-45aa-b2b9-069468f352d2.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY3OTg5NTIsIm5iZiI6MTczNjc5ODY1MiwicGF0aCI6Ii8xODU1NjU5My8zOTc5Mjk1OTItYWQ4OTI0ZTctY2EyZS00NWFhLWIyYjktMDY5NDY4ZjM1MmQyLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEzVDIwMDQxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU4YzBkMjJhNDM5ZmVkNDU4ZTk4MzQ4ZjIxMjMxZDY0YmVjNWEwMGY3ODMxODE1YWZhZjVkNjhjMmQwN2JjNDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.0TdD_MXq4CvlLD790xS__UhV_GSoIfIsyBfXV2PTwn0)
> 
> For non-canonical GPU: ![2024-12-21 21 13 30](https://private-user-images.githubusercontent.com/18556593/397929599-07fdf3e3-218c-47b5-9601-460aa3b0c1aa.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY3OTg5NTIsIm5iZiI6MTczNjc5ODY1MiwicGF0aCI6Ii8xODU1NjU5My8zOTc5Mjk1OTktMDdmZGYzZTMtMjE4Yy00N2I1LTk2MDEtNDYwYWEzYjBjMWFhLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEzVDIwMDQxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRlMDMzMTUzY2Y1YzliMzAxM2ZkNjUxMjk5NjhmMTcxNGRiYTBkNDEzMTFhZWI2NGUwNTgwNjQ1NzgzMmY3YTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.VTMahHnyDo4VgvAeYw97o2MkyJTl4-vcF06K_bYcxuU) [ ](https://private-user-images.githubusercontent.com/18556593/397929599-07fdf3e3-218c-47b5-9601-460aa3b0c1aa.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY3OTg5NTIsIm5iZiI6MTczNjc5ODY1MiwicGF0aCI6Ii8xODU1NjU5My8zOTc5Mjk1OTktMDdmZGYzZTMtMjE4Yy00N2I1LTk2MDEtNDYwYWEzYjBjMWFhLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEzVDIwMDQxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRlMDMzMTUzY2Y1YzliMzAxM2ZkNjUxMjk5NjhmMTcxNGRiYTBkNDEzMTFhZWI2NGUwNTgwNjQ1NzgzMmY3YTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.VTMahHnyDo4VgvAeYw97o2MkyJTl4-vcF06K_bYcxuU) ![2024-12-21 21 13 30](https://private-user-images.githubusercontent.com/18556593/397929599-07fdf3e3-218c-47b5-9601-460aa3b0c1aa.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY3OTg5NTIsIm5iZiI6MTczNjc5ODY1MiwicGF0aCI6Ii8xODU1NjU5My8zOTc5Mjk1OTktMDdmZGYzZTMtMjE4Yy00N2I1LTk2MDEtNDYwYWEzYjBjMWFhLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEzVDIwMDQxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRlMDMzMTUzY2Y1YzliMzAxM2ZkNjUxMjk5NjhmMTcxNGRiYTBkNDEzMTFhZWI2NGUwNTgwNjQ1NzgzMmY3YTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.VTMahHnyDo4VgvAeYw97o2MkyJTl4-vcF06K_bYcxuU) [ ](https://private-user-images.githubusercontent.com/18556593/397929599-07fdf3e3-218c-47b5-9601-460aa3b0c1aa.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY3OTg5NTIsIm5iZiI6MTczNjc5ODY1MiwicGF0aCI6Ii8xODU1NjU5My8zOTc5Mjk1OTktMDdmZGYzZTMtMjE4Yy00N2I1LTk2MDEtNDYwYWEzYjBjMWFhLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEzVDIwMDQxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRlMDMzMTUzY2Y1YzliMzAxM2ZkNjUxMjk5NjhmMTcxNGRiYTBkNDEzMTFhZWI2NGUwNTgwNjQ1NzgzMmY3YTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.VTMahHnyDo4VgvAeYw97o2MkyJTl4-vcF06K_bYcxuU)
> 
> Fine when no catalog refresh is involved, otherwise it looks a little confusing since `Optimizing...` or `List accelerators...` would last several seconds with no detailed message.

Nice finding @aylei! It seems we can solve the issue by having the detailed spinner you shown above. Let's do it in another PR. Merging this PR for now.
